# phyloscripts by mossmatters
----
Helper scripts for processing and visualization of phylogenetics datasets, mostly written in Python. For more information about each of the scripts, visit the appropriate subdirectory.


# [Alleles Workflow](https://github.com/mossmatters/phyloscripts/tree/master/alleles_workflow)

Python and shell scripts for phasing alleles from HybSeq data, as done in [Kates et al. AJB 2018](https://www.ncbi.nlm.nih.gov/pubmed/29729187).

# [PhypartsPieCharts](https://github.com/mossmatters/phyloscripts/tree/master/phypartspiecharts)

Script for plotting pie charts from bipartition analysis on a species phylogeny.


# [HybpiperUtils](https://github.com/mossmatters/phyloscripts/tree/master/hybpiperutils)

Scripts for working with target capture data recovered or assembled using [HybPiper](github.com/mossmatters/hybpiper), including `fasta_merge.py` from version 1.3.1.

# [Minority Report](https://github.com/mossmatters/phyloscripts/tree/master/minorityreport)

Display minority bipartitions on a species tree.

# [Haplonerate](https://github.com/mossmatters/phyloscripts/tree/master/haplonerate)

Adjust phased allelic sequences to have phased alleles only in the largest phase block, with ambiguity codes elsewhere. Used in [Kates et al. AJB 2018](https://www.ncbi.nlm.nih.gov/pubmed/29729187)

# Branch Length Outlier

Identify extreme branch lengths in a set of gene trees. Generates images of offending trees so that manual curation of every gene tree is not necessary.
