# Haplonerate

After using a read-backed phasing algorithm, such as GATK or WhatsHap, allelic sequences generally contain phased alleles throughout the entire reference sequence. However, there is frequently more than one phase block, especially if the data was generated with targeted sequencing (HybSeq) where multiple exons were recovered from a single locus.

In the example below, variant sites in the blue (left) block cannot be phased with alleles in the yellow (right) block.

![](img/AJB_Figure_1.pdf)

One solution is to reduce the sequence to the longest phase block, deleting other sites. This is not ideal for phylogenetic analysis, as the deleted sites may retain informative sites among species. It will also result in sequences that are of variable lengths across individuals, which may affect alignment.

Another solution would be to (hard) mask the sequence outside the longest phase block with Ns. This retains the sequence length but intra-individual informative sites are still lost.

Finally, phased alleles can be retained in the longest phase block, variant sites are replaced with with ambiguity codes in other regions. 

This script takes two files, containing phased haplotype sequences for one
or more genes and edits the sequences to retain only variable sites in the largest phase
block. The two haplotype sequences can be generated by bcftools, for example.




## Setup

1. Following variant calling, generate a new reference sequence that contains ambiguity codes. In GATK, use `FastaAlternateReferenceMaker`. 
1. Run a Read-backed phasing algorithm, such as WhatsHap (http://whatshap.readthedocs.io/), which generates a phased VCF file and a GTF file containing the locations of phase blocks.
2. Generate separate phased sequences for each gene in FASTA format, for example using `bcftools consensus` in the samtools package.
3. Run `haplonerate.py` to adjust the sequences.


## Input

**Required**: 

* GTF file annotating the locations of phase blocks.
* Two FASTA file containing sequences for one or more genes. The script assumes that the sequences are paired-- the first sequence in the first file corresponds to the first sequence in the second file.

**Options**

`--output` Specifies an output file for the edited sequences (default is `stdout`). Both alleles are written to the same file with `_h1` or `_h2` appended to the name.

`--block` Specifies a file for printing the phase block information for each gene (number of blocks, length of gene, and length of longest block).

`--edit` Which editing method is preferred. There are three options for editing the output sequences:

* **delete**: retain only the longest phase block, delete the rest of the sequence
* **ref**: use reference sequences to fill the rest of the sequence outside the longest block (default)
* **mask**: fill sequence not in the longest phase block with N

If `--edit ref` is used, the reference sequence must be supplied with `--reference`.

## Example Usages

**Default usage (`--edit ref`)**

`haplonerate.py whatshap.gtf haplotype_h1.fasta haplotype_h2.fasta --reference ambiguity_ref.fasta`

### Triploid Data

Use the `haplonerate3N.py` script instead:

`haplonerate.py whatshap.gtf haplotype_h1.fasta haplotype_h2.fasta haplotype_h3.fasta --reference ambiguity_ref.fasta --edit delete`





